### ISSUE TITLE
Checking if preview available for non HTTP applications

### ISSUE THREAD

**Issue by aure-olli** [jenkins-x/jx#7112]
### Summary 

We have gRPC applications, and we always have troubles with Jenkins-X trying the ingress in `jx preview`, using HTTP requests. Unfortunately, there is currently no way to disable this, except giving another name to the service.

I propose:
- An option to disable checking
- An option to set the protocol (HTTP, HTTPS, GRPC, GRPC-usecured, TCP, ...)
- An option to manually set the service to try
- An option to manually set the ingress to try
- Options to manually set the host, port and path to try
- Reading the `nginx.ingress.kubernetes.io/backend-protocol` annotation of the ingress to choose the protocol ([doc](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#backend-protocol))
- Trying to guess the protocol with the service port (443 -> HTTPS, 50051 -> GRPC, rest is HTTP)

### Steps to reproduce the behavior

Create a gRPC application, with a gRPC ingress (`nginx.ingress.kubernetes.io/backend-protocol: GRPC`) and make a preview.

### Expected behavior

`jx preview` should acknowledge that the gRPC server is running.

### Actual behavior

`jx preview` fails because it receives "502 Bad Gateway" from the ingress when trying HTTPS requests.

### Jx version

The output of `jx version` is:

```
NAME               VERSION
jx                 2.0.1243
Kubernetes cluster v1.14.10-gke.27
kubectl            v1.17.0
helm client        2.14.3
git                2.17.1
Operating System   Ubuntu 18.04.4 LTS
```

### Discussion

Does my solution seem reasonable ? I propose the options:
- `--skip-availability-check`
- `--availability-check-protocol` with values `HTTP`, `HTTPS`, `GRPC`, `GRPC-unsecured` and `TCP`
- `--availability-check-service`
- `--availability-check-ingress`
- `--availability-check-host`
- `--availability-check-port`
- `--availability-check-path`

**Comment by aure-olli** [comment]
@aure-olli,  nice one. Makes sense. In the current situation, a `skip-availability-check` makes sense, since it is not a given that the "exposed service" is HTTP based. 

Would you mind creating an issue for just this PR and the `skip-availability-check` flag? The new issue could also reference #7112 for the bigger picture. 

Regarding #7112, I am wondering whether we need to discuss first a bit more about the expectations of a "preview". ATM, it is pretty much an exposed HTTP(S) service. I am wondering whether the assumption is not a bit limiting. WDYT?

_Originally posted by @hferentschik in https://github.com/jenkins-x/jx/pull/7143#issuecomment-623972267_

**Comment by aure-olli** [comment]
Our current usage of the preview is to test any update of our applications. Some applications are HTTP frontends, some are HTTP backends, some are gRPC backends, some are python libraries, some kubernetes tools.

- Backends would usually spawn their own database, and a local deployment of the frontend. Since `jx preview` automatically finds the ingress, it would usually find the backend's ingress, which is fine in this case. Having a backend preview also helps to develop the frontend in parallel, because we would configure the frontend's preview to use the backend's preview.

- For gRPC backends, we usually don't have any frontend, we test it by code. But even if we had a frontend, `jx preview` would probably find the backend's ingress and fail. The preview creates an URL for us to test our new backend easily.

- For python libraries, we often have a minimal frontend to test it a little, but otherwise there is no ingress nor server so `jx preview` succeeds. But the main point is for the pipeline to create us a pipy package with a special version number, than we can then install in out preview backends and test.

- For the kubernetes tools, there is no ingress nor server so `jx preview` succeeds. We configure our tools to recognize different annotations' prefix from the production one, so we can test without any interferance.

In general, we find preview really useful to test anything in a specific and isolated environment. Deploy a preview, give the hostname / IP / database access / package version to anybody to try it, then approve when you're statisfied. 

Our usage is unrelated to HTTP. The "availability check" is however useful to know when it is ready to test. But since deployments have a `readinessProbe` already, can't we just in general wait for it ? Or even wait for everything to be ready in the environment ?
